src/components/Form.js

```diff
@@ -0,0 +1,214 @@
+    formID: PropTypes.string.isRequired,
+
+    /** Text to be displayed in the submit button */
+    submitButtonText: PropTypes.string.isRequired,
+
+    /** Callback to validate the form */
+    validate: PropTypes.func.isRequired,
+
+    /** Callback to submit the form */
+    onSubmit: PropTypes.func.isRequired,
+
+    children: PropTypes.node.isRequired,
+
+    /* Onyx Props */
+
+    /** Contains the form state that must be accessed outside of the component */
+    formState: PropTypes.shape({
+
+        /** Controls the loading state of the form */
+        isSubmitting: PropTypes.bool,
+
+        /** Server side error message */
+        serverErrorMessage: PropTypes.string,
+    }),
+
+    // eslint-disable-next-line react/forbid-prop-types
+    draftValues: PropTypes.object,
+
+    ...withLocalizePropTypes,
+};
+
+const defaultProps = {
+    formState: {
+        isSubmitting: false,
+        serverErrorMessage: '',
+    },
+    draftValues: {},
+};
+
+class Form extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            errors: {},
+        };
+
+        this.inputRefs = {};
+        this.touchedInputs = {};
+
+        this.getValues = this.getValues.bind(this);
+        this.setTouchedInput = this.setTouchedInput.bind(this);
+        this.validate = this.validate.bind(this);
+        this.submit = this.submit.bind(this);
+    }
+
+    /**
+     * @param {String} inputID - The inputID of the input being touched
+     */
+    setTouchedInput(inputID) {
+        this.touchedInputs[inputID] = true;
+    }
+
+    /**
+     * @returns {Object} - An object containing the values for each inputID
+     */
+    getValues() {
+        const values = {};
+        _.each(this.inputRefs, (inputRef, inputID) => {
+            values[inputID] = inputRef.value;
+        });
+        return values;
+    }
+
+    submit() {
+        // Return early if the form is already submitting to avoid duplicate submission
+        if (this.props.formState.isSubmitting) {
+            return;
+        }
+
+        // Touches all form inputs so we can validate the entire form
+        _.each(this.inputRefs, (inputRef, inputID) => (
+            this.touchedInputs[inputID] = true
+        ));
+
+        // Validate form and return early if any errors are found
+        const values = this.getValues();
+        if (!_.isEmpty(this.validate(values))) {
+            return;
+        }
+
+        // Set loading state and call submit handler
+        FormActions.setIsSubmitting(this.props.formID, true);
+        this.props.onSubmit(values);
+    }
+
+    /**
+     * @param {Object} values - An object containing the value of each inputID
+     * @returns {Object} - An object containing the errors for each inputID
+     */
+    validate(values) {
+        FormActions.setServerErrorMessage(this.props.formID, '');
+        const validationErrors = this.props.validate(values);
```

Comment:
This seems tightly coupled with the implementation of the `validate` prop, and the expected behavior isn't well documented. It takes an object like `{inputID1: value1, inputID2: value2}`, and returns an object like `{inputID1: error1, inputID2: error2}` ?

We should find a way to document/enforce that return type. Can we get fancy with propTypes and validate the return type of the `validate` function? Maybe throw an error if it doesn't return the expected type?

Reply:
Someone could hypothetically just look at usages of the `validate()` prop and see how it is used. I think we are planning to write some kind of markdown doc by the end of this that explains everything and the overall philosophy WRT forms.

Really like the idea to throw if we get anything other than an object here. Then again, maybe there is some case where you don't want to validate the input at all, and then `this.props.validate` should default to a function that returns an empty object.

Reply:
>We should find a way to document/enforce that return type

Like the spirit of this, but empty object is a possible return type - so enforcing anything else seems hard.

Reply:
I added an example to the validate prop docs in Form. As Marc mentioned, an empty object is a valid return and would be quite common if we found no errors when validating. I think it might be quite hard to enforce this, so I think that we should probably rely on docs/examples.

Reply:
Got it, well we can throw a meaningful error if `this.props.validate` doesn't return an object (empty or not):

```js
const validationErrors = this.props.validate(values);
if (!_.isObject(validationErrors) || _.isArray(validationErrors) || _.isFunction(validationErrors) {
    throw new Error('Form validate prop must take an object with shape {inputID: value}, and return an object with shape {inputID: error}');
}
```

That's something 🤷 Demoting this to NAB if others don't see it as valuable.

Reply:
Added a check in `validate`. Thanks @roryabraham!

Reply:
I like that error message @roryabraham and think it would be valuable to add.